### PR TITLE
Initial commit with CI file

### PR DESCRIPTION
**Pull request by toninis**
#### Summary

Migrate CI/CD from CircleCI to Github Actions 

> Used Golang 1.19 since tests are broken with 1.18 . Check [here](https://github.com/mattermost/mattermost-plugin-apps/actions/runs/4353420476/jobs/7607437928). I am open for suggestions here

#### Ticket Link

https://mattermost.atlassian.net/browse/CLD-5248

### CODE REVIEW

**Review by phoinixgrr — APPROVED** [review]

File: .github/workflows/cd.yaml [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  workflow_run:
+    workflows:
+      - ci
+    types:
+      - completed
+    branches:
+      - master
+
+env:
+  TERM: xterm
```
Comment (spirosoik) [inline .github/workflows/cd.yaml:16]: is there a requirement for this?

File: .github/workflows/cd.yaml [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+      - name: ci/checkout-repo
+        uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
+
+      - name: ci/setup-go
+        uses: actions/setup-go@6edd4406fa81c3da01a34fa6f6343087c207a568 # v3.5.0
+        with:
+          go-version: "${{ env.GO_VERSION }}"
+          cache: true
+
+      - name: ci/ensure-build-on-all-platforms
```
Comment (spirosoik) [inline .github/workflows/cd.yaml:34]: let's call just `ci/build`

File: .github/workflows/cd.yaml [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        run: make dist
+
+      - name: ci/generate-release-notes
```
Comment (spirosoik) [inline .github/workflows/cd.yaml:37]: I have seen this multiple times in different repos. Let's mark this to be an action or even do it right now.

File: .github/workflows/cd.yaml [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        run: |
+          printf "Supported Mattermost Server Versions: **$(cat plugin.json | jq .min_server_version -r)+** \n## Enhancements\n\n## Fixes\n"  >> dist/release-notes.md
+          if [[ $(git tag -l | wc -l) -eq 1 ]]; then
+            git log --pretty='format:- %h %s' --abbrev-commit --no-decorate --no-color $(git rev-list --max-parents=0 HEAD) HEAD >> dist/release-notes.md
+          else
+            git log --pretty='format:- %h %s' --abbrev-commit --no-decorate --no-color $(git describe --tags --abbrev=0 $(git describe --tags --abbrev=0)^)..HEAD >> dist/release-notes.md
+          fi
+
+      - name: ci/persist-build-artifacts
```
Comment (spirosoik) [inline .github/workflows/cd.yaml:46]: `ci/upload-artifacts`, we need to revisit these names to have a unify thing so we can understand the same across all CIs

File: .github/workflows/cd.yaml [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+      - name: cd/download-artifacts
+        uses: actions/download-artifact@9bc31d5ccc31df68ecc42ccf4149144866c47d8a # v3.0.2
+        with:
+          name: dist
+          path: dist
+
+      - name: cd/rename-artifacts
+        run: mv dist/*.tar.gz dist/${GITHUB_REPOSITORY#*/}-ci.tar.gz
+
+      - name: cd/rename-artifacts
```
Comment (spirosoik) [inline .github/workflows/cd.yaml:73]:
```suggestion
      - name: cd/setup-aws
```

File: .github/workflows/ci.yaml [hunk 6]
```diff
@@ -0,0 +1,418 @@
+name: ci
+on:
+  pull_request:
+  push:
+    branches:
+      - master
+
+env:
+  TERM: xterm
```
Comment (spirosoik) [inline .github/workflows/ci.yaml:9]: same here

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+        with:
+          go-version: "${{ env.GO_VERSION }}"
+          cache: true
+
+      - name: ci/ensure-build-on-all-platforms
+        run: make dist
+
+      - name: ci/generate-release-notes
+        run: |
+          printf "Supported Mattermost Server Versions: **$(cat plugin.json | jq .min_server_version -r)+** \n## Enhancements\n\n## Fixes\n"  >> dist/release-notes.md
```
Comment (spirosoik) [inline .github/workflows/ci.yaml:70]: again same here. I think it worth to make it an action

File: .github/workflows/ci.yaml [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+          if [[ $(git tag -l | wc -l) -eq 1 ]]; then
+            git log --pretty='format:- %h %s' --abbrev-commit --no-decorate --no-color $(git rev-list --max-parents=0 HEAD) HEAD >> dist/release-notes.md
+          else
+            git log --pretty='format:- %h %s' --abbrev-commit --no-decorate --no-color $(git describe --tags --abbrev=0 $(git describe --tags --abbrev=0)^)..HEAD >> dist/release-notes.md
+          fi
+
+      - name: ci/persist-build-artifacts
```
Comment (spirosoik) [inline .github/workflows/ci.yaml:77]: naming

File: .github/workflows/nightly.yaml [hunk 9]
```diff
@@ -0,0 +1,356 @@
+name: nightly
```
Comment (spirosoik) [inline .github/workflows/nightly.yaml:1]: is it my impression or we do the same things with the `ci.yaml` right? Why we are not adding the schedule on `ci.yaml`  and it will run when is needed? We did a similar thing to plugins

**Review by spirosoik — CHANGES_REQUESTED** [review]
great job overall. a few comments to discuss

File: .github/workflows/cd.yaml (hunk 1 above)
Comment (toninis) [reply .github/workflows/cd.yaml:16]: Yep . Makefile needs that because it has some erroneous errors

File: .github/workflows/nightly.yaml (hunk 9 above)
Comment (toninis) [reply .github/workflows/nightly.yaml:1]: Yep . Plus we are running e2e tests and we do not run coverage / dist . I implemented this in the first place but I did not like the conditions . I will merge them though you are right

Comment (toninis) [reply .github/workflows/nightly.yaml:1]: Actually another reason I 've separated those is to easily distinguish them from the Github Web UI . If we merge them we will have to distinguish them with filters

File: build/custom.mk [hunk 10]
```diff
@@ -28,7 +28,6 @@ ifneq ($(HAS_SERVER),)
 	mockgen -destination server/mocks/mock_upstream/mock_upstream.go github.com/mattermost/mattermost-plugin-apps/upstream Upstream
 	mockgen -destination server/mocks/mock_store/mock_appstore.go github.com/mattermost/mattermost-plugin-apps/server/store AppStore
 	mockgen -destination server/mocks/mock_store/mock_session.go github.com/mattermost/mattermost-plugin-apps/server/store SessionStore
-	mockgen -destination server/mocks/mock_store/mock_app.go github.com/mattermost/mattermost-plugin-apps/server/store AppStore
```
Comment (mickmister) [inline build/custom.mk:31]: Is this change necessary?

File: Makefile [hunk 11]
```diff
@@ -46,13 +51,8 @@ ifneq ($(HAS_WEBAPP),)
...
 
 ifneq ($(HAS_SERVER),)
-	@if ! [ -x "$$(command -v golangci-lint)" ]; then \
-		echo "golangci-lint is not installed. Please see https://github.com/golangci/golangci-lint#install for installation instructions."; \
-		exit 1; \
-	fi; \
-
 	@echo Running golangci-lint
-	golangci-lint run ./...
+	golangci-lint run ./... --verbose
```
Comment (mickmister) [inline Makefile:55]: `--verbose` can make CI take longer to run. Are we sure we want this to run this way in CI?

File: build/custom.mk (hunk 10 above)
Comment (toninis) [reply build/custom.mk:31]: It's a duplicate from line 19 with a different destination file

File: Makefile (hunk 11 above)
Comment (toninis) [reply Makefile:55]: It only takes some seconds to run and it's nice to have the verbosity in place

**Review by mickmister — APPROVED** [review]

**Comment by mickmister** [comment]
This is a lot of configuration that will be duplicated across several plugin I presume. Is there a way we can be more DRY and share the bulk of this in a common repo?

**Comment by toninis** [comment]
> This is a lot of configuration that will be duplicated across several plugin I presume. Is there a way we can be more DRY and share the bulk of this in a common repo?

This is part of our work . We are going to create reusable actions across the company for consistency . We will get there 😄 
Feel free thought o experiment .
